### ISSUE TITLE
Automatically adding new catalogs to the workspace, when installing a new dependency

### ISSUE THREAD

**Issue by zkochan** [pnpm/pnpm#8876]
It would be great if we could also have the option to allow for all installs to have their versions managed in the default catalog  (if one exists). Meaning that it would not only use the `catalog:` when an existing version of the same package is in the default catalog, but also if there is no existing entry, rather than adding the version to the `package.json`, it adds the package version to the catalog, and adds the `catalog:` reference to `package.json` in that workspace.

I can understand if this is not the desired default behavior for all, but would be great if we could have some config value in `.npmrc` to enable this or possible with a install flag (eg. `pnpm add <pkg> --catalog`)

_Originally posted by @nickfujita in https://github.com/pnpm/pnpm/issues/8640#issuecomment-2431394664_

**Comment by martinkors** [comment]
> It seems like a distinction between two settings might be useful:
> 
> `preferCatalog=true` seems to me to suggest using the catalog if not specified and no conflict of version results. Otherwise, just install the literal version like you said. (Maybe make a note of what happened eg: "Conflicted with catalog version so literal version installed in package" or something)
> 
> `requireCatalog=true` would only ever install to the catalog. Behaves like preferCatalog except in the case that a version is specified that doesn't match catalog, could then error. Could even suggest that the user create a named catalog if a truly different version of the dep was required
> 
> In both cases you could offer the user a prompt / choice "Version doesn't match catalog and requireCatalog / preferCatalog is true. Update catalog version? (will affect  all consuming packages)?" or similar

Thanks for the incredible feedback!

I really love the distinction you give between two possible use cases. I believe both would be beneficial to have. Some developers would probably like to use it loosely, while some might want it to be strict.

I propose that both options be available through a .npmrc field called `useCatalog` with the possible values being something like `strict | loose | manual`.

Where `strict` behaves as your `requireCatalog`, `loose` being to use catalog on version match or default (with a warning if not matched) and lastly `manual` referring to the default behavior of using the `--catalog` argument.

**Comment by scamden** [comment]
That’s smart since they are basically mutually exclusive options. One idea for names

use-catalog: always | prefer | manual

**Comment by martinkors** [comment]
> That’s smart since they are basically mutually exclusive options. One idea for names
> 
> use-catalog: always | prefer | manual

Great names. I will try to get it working as soon as possible, but it might take a few days since I am limited in terms of time.

**Comment by gluxon** [comment]
Strongly agree with @scamden that this feature is tied to how users expect catalogs to be enforced in their repos.

- https://github.com/pnpm/pnpm/issues/8644

Ideally adding a new catalog entry (or updating) oone doesn't immediately break checks the repo has configured.

@martinkors I appreciate the help! Would you be able to pause on implementation for a bit while I finish a few thoughts on how catalog checks would be configured? I'll try to finish thinking through it this week. I think that will affect how this is implemented.

Separately, one of the challenges of adding new entries to the catalog is that pnpm doesn't currently use a white-space preserving yaml serializer. It would be ideal to complete that work before muting the `pnpm-workspace.yaml` file destructively.

**Comment by martinkors** [comment]
> @martinkors I appreciate the help! Would you be able to pause on implementation for a bit while I finish a few thoughts on how catalog checks would be configured? I'll try to finish thinking through it this week. I think that will affect how this is implemented.

@gluxon of course I can pause implementation of this, no problem.

> Separately, one of the challenges of adding new entries to the catalog is that pnpm doesn't currently use a white-space preserving yaml serializer. It would be ideal to complete that work before muting the `pnpm-workspace.yaml` file destructively.

I have been trying to modify the currently used yaml serializer to preserve comments and whitespaces and I believe it is feasible to do so. Still haven't gotten this 100% working but it currently preserves comments but not whitespaces.

**Comment by fpapado** [comment]
@martinkors, if it is any help, what I did when implementing catalog upgrades in Renovate, was use a Concrete Syntax Tree (CST) representation, in order to help preserve whitespace, comments, and all the odd YAML combinations of syntax. Or rather, I hope the CST handles all of the combinations; I verified many with tests 😌 

The underlying code is at https://github.com/renovatebot/renovate/blob/8d7e9eb967d2a160633431dbdacca190e1e83b99/lib/modules/manager/npm/update/dependency/pnpm.ts#L38-L45, but the core idea is using [the `yaml` library off npm](https://www.npmjs.com/package/yaml), [which spells out the syntax preservation as a goal of its CST](https://eemeli.org/yaml/#working-with-cst-tokens):

> For most use cases, the Document or pure JS interfaces provided by the library are the right tool. Sometimes, though, it's important to keep the original YAML source in as pristine a condition as possible. For those cases, the concrete syntax tree (CST) representation is provided, as it retains every character of the input, including whitespace.

**Comment by martinkors** [comment]
> Heads up: there's some earlier discussion about the YAML libraries in pnpm (`js-yaml`, the current one, vs `yaml`) in https://github.com/pnpm/pnpm/pull/6273#discussion_r1431894749 and related threads. The calculus for sticking with `js-yaml` might be different now that we are talking about `pnpm-workspace.yaml` files, which are modified directly by users and also by pnpm. This is in contrast to `pnpm-lock.yaml`, which pnpm can be the sole writer of (and thus comment/syntax preservation could be considered more of a nice-to-have).

Yes I was aware of this, but personally I think wanting to switch from `js-yaml` to `yaml` is a bit odd, as `js-yaml` from my experience performs way better than `yaml`.

I don't know which direction is best since they each have their own pros and cons.

**Comment by martinkors** [comment]
@gluxon what is the current status of this?

I saw @zkochan mention that comment preservation shouldn't be a blocker in [this issue](https://github.com/pnpm/pnpm/issues/8641#issuecomment-2753980944).

I would love to finalize my work for this issue.

**Comment by gluxon** [comment]
@martinkors Sorry for the holdup. I wanted to think through how this would work under a new mechanism that could restrict certain packages to using certain named catalogs, but I think there's enough complexity there that we should move forward with this issue. I can figure out the constraint mechanism later if we decide to do it.

A heads up that there's a similar PR opened recently to add a `--save-catalog` field. I think we'll still want to followup to that PR and make this an option.

https://github.com/pnpm/pnpm/pull/9484

**Comment by martinkors** [comment]
> [@martinkors](https://github.com/martinkors) Sorry for the holdup. I wanted to think through how this would work under a new mechanism that could restrict certain packages to using certain named catalogs, but I think there's enough complexity there that we should move forward with this issue. I can figure out the constraint mechanism later if we decide to do it.

No problem at all! 

During this time I have gotten some more experience with the catalogs by converting my team's big projects to using it. Using catalogs is currently still a little rough due to some of the commands like `pnpm add` and `pnpm update` still missing support for it, but it has made keeping track of dependencies a lot easier.

One thing I have noticed though is that some sort of bulk moving from one named/default catalog to another based on some rules, would be extremely useful for getting existing projects converted to using catalogs in a way that is manageable.

The same story also applies to this PR in particular. I think it could be beneficial to have some sort of regex or glob rules for sorting dependencies into specific catalogs. The solution used for this would probably also be useful when using the `codemod`.

These rules/filters could be located either in `.npmrc` as 
```
catalogRule:myCatalog1="@uilib1/*"
catalogRule:myCatalog2="/@uilib2\\/.+/"
``` 
or in `pnpm-workspace.yaml` as
```
catalogRules:
  myCatalog1: '@uilib1/*'
  myCatalog2: /@uilib2\/.+/
```

> A heads up that there's a similar PR opened recently to add a `--save-catalog` field. I think we'll still want to followup to that PR and make this an option.
> 
> * [feat: `pnpm add` option to add new entries to catalogs  #9484](https://github.com/pnpm/pnpm/pull/9484)

Thanks for the heads up, the work in #9484 looks really good!

**Comment by zkochan** [comment]
What makes it useful to group catalogs by package name? I thought the main point of named catalogs was to allow "version groups". When you need react16 and react17, for instance.

**Comment by jycouet** [comment]
My main usecase is to have ONE place (mono repo with 7 package.json)

And I like to group by feature "eslint", "..." Adding also comments around.

Then I also have a few calalogs for versions when needed.

**Comment by martinkors** [comment]
I use catalogs both for managing different versions and also as a good overview of my dependencies. 
An example could be the following:
```yaml
catalogs:
  eslint:
    eslint: x.x.x
    myeslintconfig: x.x.x
    my-eslint-plugin: x.x.x

  ui:
    '@my-ui/button': x.x.x
    '@my-ui/switch': x.x.x
    '@my-ui/icon': x.x.x
    '@my-ui/tooltip': x.x.x
    '@my-ui/menu': x.x.x
    '@my-ui/animation': x.x.x

  react18:
    react: 18.x.x
    myoldcomponentlib: x.x.x


  react19:
    react: 19.x.x
    mynewcomponentlib: x.x.x
```

**Comment by zigang93** [comment]
Can we use -C as a shorthand for --save-catalog-name?
